core/iwasm/interpreter/wasm_runtime.c

```diff
@@ -162,10 +162,10 @@ memory_instantiate(WASMModuleInstance *module_inst, WASMModuleInstance *parent,
                    char *error_buf, uint32 error_buf_size)
 {
     WASMModule *module = module_inst->module;
-    uint64 memory_data_size, max_memory_data_size;
-    uint32 heap_offset = num_bytes_per_page * init_page_count;
-    uint32 inc_page_count, aux_heap_base, global_idx;
+    uint32 inc_page_count, global_idx;
     uint32 bytes_of_last_page, bytes_to_page_end;
+    uint64 aux_heap_base, heap_offset = num_bytes_per_page * init_page_count;
```

Comment:
Use `heap_offset = (uint64)num_bytes_per_page * init_page_count` to avoid possible integer overflow, suppose num_bytes_per_page and init_page_count are both 65536?

Reply:
And similar issue occurs in L212 and L254, how about disallowing insert app heap when initial memory size is 4G? For example, add the check before L209?

And similar issues may occur in aot_runtime.c.